lib/spack/spack/solver/asp.py

```diff
@@ -3829,8 +3829,10 @@ def execute_explicit_splices(self):
         for splice_set in splice_config:
             target = splice_set["target"]
             replacement = spack.spec.Spec(splice_set["replacement"])
-            assert replacement.abstract_hash
-            replacement.replace_hash()
+
+            if not replacement.abstract_hash:
+                raise ValueError("explicit splice replacement must be specified by hash")
```

Comment:
I don't think ValueError is the right thing here... really should be some type of error defined in asp.py.  Maybe `SolverError`?  I don't like it that it extends `InternalConcretizerError`, but it's the right name.

This should also report what line in the config the error happened at. You have `splice_set["replacement"]`, which should have the right annotations.